src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,533 @@
+{
+	unsigned long accum = 0, *pte = NULL;
+
+	assert(pt);
+	assert((PGTBL_FLAG_MASK & addr) == 0);
+
+	return __pgtbl_expandn(pt, addr >> PGTBL_PAGEIDX_SHIFT, PGTBL_DEPTH + 1, &accum, &pte, NULL);
+}
+
+int
+chal_pgtbl_mapping_scan(struct cap_pgtbl *pt)
+{
+	unsigned int i, pte, *page;
+	livenessid_t lid;
+	u64_t        past_ts;
+
+	/**
+         * This scans the leaf level of the pgtbl and verifies
+	 * quiescence.
+         */
+	if (pt->lvl != PGTBL_DEPTH - 1) return -EINVAL;
+
+	page = (unsigned int *)(pt->pgtbl);
+	assert(page);
+
+	for (i = 0; i < PAGE_SIZE / sizeof(int *); i++) {
+		pte = *(page + i);
+		if (pte & X86_PRESENT || pte & X86_COSFRAME) return -EINVAL;
+
+		if (pte & X86_QUIESCENCE) {
+			lid = pte >> PGTBL_PAGEIDX_SHIFT;
+
+			if (ltbl_get_timestamp(lid, &past_ts)) return -EFAULT;
+			if (!tlb_quiescence_check(past_ts)) return -EQUIESCENCE;
+		}
+	}
+
+	return 0;
+}
+
+void *
+chal_pgtbl_lkup_lvl(pgtbl_t pt, u32_t addr, u32_t *flags, u32_t start_lvl, u32_t end_lvl)
+{
+	return __pgtbl_lkupani((pgtbl_t)((unsigned long)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT, start_lvl,
+	                       end_lvl, flags);
+}
+
+int
+chal_pgtbl_ispresent(u32_t flags)
+{
+	return flags & (X86_PRESENT | X86_COSFRAME);
+}
+
+unsigned long *
+chal_pgtbl_lkup(pgtbl_t pt, u32_t addr, u32_t *flags)
+{
+	void *ret;
+
+	ret = __pgtbl_lkupan((pgtbl_t)((unsigned long)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT, PGTBL_DEPTH + 1,
+	                     flags);
+	if (!pgtbl_ispresent(*flags)) return NULL;
+	return ret;
+}
+
+unsigned long *
+chal_pgtbl_lkup_pte(pgtbl_t pt, u32_t addr, u32_t *flags)
+{
+	return __pgtbl_lkupan((pgtbl_t)((unsigned long)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT, PGTBL_DEPTH,
+	                      flags);
+}
+
+int
+chal_pgtbl_get_cosframe(pgtbl_t pt, vaddr_t frame_addr, paddr_t *cosframe)
+{
+	u32_t          flags;
+	unsigned long *pte;
+	paddr_t        v;
+
+	pte = pgtbl_lkup_pte(pt, frame_addr, &flags);
+	if (!pte) return -EINVAL;
+
+	v = *pte;
+	if (!(v & X86_COSFRAME)) return -EINVAL;
+
+	*cosframe = v & PGTBL_FRAME_MASK;
+
+	return 0;
+}
+
+pgtbl_t
+chal_pgtbl_create(void *page, void *curr_pgtbl)
+{
+	pgtbl_t ret = pgtbl_alloc(page);
+	/* Copying the kernel part of the pgd. */
+	memcpy(page + KERNEL_PGD_REGION_OFFSET, (void *)chal_pa2va((paddr_t)curr_pgtbl) + KERNEL_PGD_REGION_OFFSET,
+	       KERNEL_PGD_REGION_SIZE);
+
+	return ret;
+}
+
+int
+chal_pgtbl_quie_check(u32_t orig_v)
+{
+	livenessid_t lid;
+	u64_t        ts;
+
+	if (orig_v & X86_QUIESCENCE) {
+		lid = orig_v >> PGTBL_PAGEIDX_SHIFT;
+		/**
+                 * An unmap happened at this vaddr before. We need to
+		 * make sure that all cores have done tlb flush before
+		 * creating new mapping.
+                 */
+		assert(lid < LTBL_ENTS);
+
+		if (ltbl_get_timestamp(lid, &ts)) return -EFAULT;
+		if (!tlb_quiescence_check(ts)) {
+			printk("kern tsc %llu, lid %d, last flush %llu\n", ts, lid,
```

Comment:
I have not touched Quiescence so I have no idea if it is important to have this printk. I noticed the use of `printk` before in the same PR for a Quiescence related code I think.

Reply:
should not be printks in the kernel unless it is bombing out.